impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,200 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
+     * needed :(
+     *
      * @param principals
      * @return
      */
     public List<String> getGroupsFromPrincipals(Iterable<Principal> principals) {
         List<String> groups = new ArrayList<>();
- 
+
         for (Principal principal : principals) {
             if (principalToGroups(principal, groups)) {
                 // return value of true means we're done early. This can be used
                 // when we know there's only 1 principal holding all the groups
                 return groups;
             }
         }
- 
+
         return groups;
     }
-     
+
     public List<String> principalToGroups(Principal principal) {
         List<String> groups = new ArrayList<>();
         principalToGroups(principal, groups);
         return groups;
     }
-    
+
     private Principal doGetCallerPrincipalFromPrincipals(Iterable<Principal> principals) {
-        
         for (Principal principal : principals) {
-            switch (principal.getClass().getName()) {
-                case "org.glassfish.security.common.PrincipalImpl": // GlassFish/Payara
-                // JBoss EAP/WildFly convention 1 - single top level principal of the below type
-                case "org.jboss.security.SimplePrincipal":
-                case "com.ibm.ws.security.authentication.principals.WSPrincipal": // Liberty
-                    return principal;
-                // JBoss EAP/WildFly convention 2 - the one and only principal in group called CallerPrincipal
-                case "org.jboss.security.SimpleGroup": 
-                    if (principal.getName().equals("CallerPrincipal") && principal instanceof Group) {
-                        
-                        Enumeration<? extends Principal> groupMembers = ((Group) principal).members();
-                        
-                        if (groupMembers.hasMoreElements()) {
-                            return groupMembers.nextElement();
-                        }
-                    }
-                    break;
-                case "org.apache.tomee.catalina.TomcatSecurityService$TomcatUser": // TomEE
-                    try {
-                        return 
-                            (Principal) Class.forName("org.apache.catalina.realm.GenericPrincipal")
-                                             .getMethod("getUserPrincipal")
-                                             .invoke(
-                                                 Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
-                                                      .getMethod("getTomcatPrincipal")
-                                                      .invoke(principal))
-                            
-                            ;
-                        } catch (Exception e) {
-                            
-                        }
-                    break;
-            }
-
+            // If the principal is a standard one, return it
```

Comment:
I don't think we can do this first -- the presence of this type doesn't necessasarily imply that it's "the" caller principal, it may simply be "a" caller principal. This would only be present on WebLogic, for example, if it were added to the Subject as an application principal -- and given the signature of notifyContainerAboutLogin(CredentialValidationResult), that could easily happen.